### PR TITLE
Autorequire luatest in server instances

### LINKED ISSUE #233: Autorequire luatest in all instances to simplify `:exec()` calls

### PR DESCRIPTION
**Pull request by ochaplashkin**
Luatest module available without explicit require:

```lua
g.server:exec(function()
   t.assert(...)
end)
```

Resolves #233

- [x] Tests
- [x] Changelog
- [x] Documentation

### CODE REVIEW

**Comment by ochaplashkin** [comment]
## The process of solving the issue

### 1. Quick solution

Added the variable `t` to the `server_instance.lua` file

```lua
-- server_instance.lua
...
box.cfg(box_cfg())
box.schema.user.grant('guest', 'super', nil, nil, {if_not_exists = true})

t = require('luatest')
...
```
**Error:** this only works if the server is created based on the server_instance.lua. Other (clean, custom) servers are not working.

### 2. Changing the server instance 

```lua
-- server.lua
function Server:exec(fn, args, options)
  ...
  local t = require('luatest')
  return exec_tail(self:eval([[
          local dump, args = ...
          local fn = loadstring(dump)
  ...
```

**Error:** incorrect scope of the `t` variable.

### 3. Declaring a local variable on the server

```lua
-- server.lua
function Server:exec(fn, args, options)
  ...
  return exec_tail(self.net_box:eval([[
          local t = require('luatest')
          local dump, args = ...
          local fn = loadstring(dump)
  ...
```
**Error:** upvalues exception in lua

### 4. Declaring a global variable on the server

```lua
-- server.lua
function Server:exec(fn, args, options)
  ...
  return exec_tail(self.net_box:eval([[
          local t = require('luatest')
          rawset(_G, 't', _t)
          local dump, args = ...
          local fn = loadstring(dump)
  ...
```
**Error:** if the environment variable `LUA_PATH` is not set, then nothing works (and does not say about it)

### 5. Added import via call and logging

```lua
-- server.lua
function Server:exec(fn, args, options)
  ...
  return exec_tail(self.net_box:eval([[
      local is_ok, _t = pcall(require, 'luatest')
      if is_ok then
            rawset(_G, 't', _t)
        else
            require('log').warn('LUA_PATH is unset, ' .. _t)
        end
     local dump, args = ...
     local fn = loadstring(dump)
     ...
     ]]))
end
```

It looks fine. However, we noticed that the `exec` method uses `eval`.

### Final. Move our solution to the `eval` method

```lua
--server.lua
function Server:eval(...)
    ...
    self.net_box:eval([[
        local is_ok, _t = pcall(require, 'luatest')
        if is_ok then
            rawset(_G, 't', _t)
        else
            require('log').warn('LUA_PATH is unset, ' .. _t)
        end
        ]]
    )
    return self.net_box:eval(...)
end
```

## Usage

```lua
-- work with single instance
server:exec(function() t.assert_equals(1,1) end)
server:eval("t.assert_equals(1,1)")

-- work with cluster
replica:exec(function() t.assert_equals(1,1) end)
replica:eval("t.assert_equals(1,1)")

```

If the `LUA_PATH` variable is not set, then you will see a warning on the server:
```
server | 2023-01-10 16:03:36.457 [19289] main/109/main eval:5 W> LUA_PATH is unset, module 'luatest' not found:
```

File: luatest/server.lua [hunk 1]
```diff
@@ -15,6 +15,7 @@ local net_box = require('net.box')
 local uri = require('uri')
 local _, luacov_runner = pcall(require, 'luacov.runner') -- luacov may not be installed
 
+
```
Comment (ylobankov) [inline luatest/server.lua:18]: Unnecessary change.

File: luatest/server.lua [hunk 2]
```diff
@@ -518,7 +528,7 @@ end
 --    -- sum == 3
 --
 --    server:exec(function()
---        local t = require('luatest')
+--        -- luatest is always available as t
```
Comment (ylobankov) [inline luatest/server.lua:531]: t -> \`t\`

File: luatest/server.lua [hunk 3]
```diff
@@ -537,8 +547,7 @@ function Server:exec(fn, args, options)
         )
         error(err, 2)
     end
-
-    return exec_tail(self.net_box:eval([[
+    return exec_tail(self:eval([[
```
Comment (ylobankov) [inline luatest/server.lua:550]: Please leave empty line as it was before.

File: luatest/server.lua (hunk 1 above)
Comment (ochaplashkin) [reply luatest/server.lua]: Removed.

File: luatest/server.lua (hunk 2 above)
Comment (ochaplashkin) [reply luatest/server.lua]: Done.

File: luatest/server.lua (hunk 3 above)
Comment (ochaplashkin) [reply luatest/server.lua:550]: Yep, added

File: luatest/server.lua [hunk 4]
```diff
@@ -480,6 +480,17 @@ function Server:eval(...)
     if self.net_box == nil then
         error('net_box is not connected', 2)
     end
+    self.net_box:eval([[
```
Comment (ylobankov) [inline luatest/server.lua:483]: This function may look like this:

```lua
function Server:eval(code, ...)
    if self.net_box == nil then
        error('net_box is not connected', 2)
    end
    local preamble = [[
        if not rawget(_G, 't') then
            local is_ok, _t = pcall(require, 'luatest')
            if is_ok then
                rawset(_G, 't', _t)
            else
                require('log').warn('LUA_PATH is unset or incorrect, ' .. _t)
            end
        end
    ]]
    return self.net_box:eval(preamble .. code, ...)
end
```

In such a way, we will have only one `self.net_box:eval` call.

File: test/autorequire_luatest_test.lua [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+g.test_exec_with_local_duplicate = function()
+    g.server:exec(function()
+        local tt = require('luatest')
+        t.assert_equals(1, 1)
+        tt.assert_equals(1, 1)
+        t.assert_equals(tt, t)
+    end)
+end
+
+g.test_eval_with_asserts = function()
```
Comment (ylobankov) [inline test/autorequire_luatest_test.lua:72]: What about `test_eval_with_t`?

File: test/autorequire_luatest_test.lua [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+    g.server:connect_net_box()
+end)
+
+g.after_all(function()
+    g.server:stop()
+    fio.rmtree(datadir)
+end)
+
+g.test_exec_without_asserts = function()
```
Comment (ylobankov) [inline test/autorequire_luatest_test.lua:41]: What about `test_exec_without_t`?

File: luatest/server.lua (hunk 4 above)
Comment (ochaplashkin) [reply luatest/server.lua]: Done. Thanks!

File: test/autorequire_luatest_test.lua (hunk 5 above)
Comment (ochaplashkin) [reply test/autorequire_luatest_test.lua]: Fixed.

File: test/autorequire_luatest_test.lua (hunk 6 above)
Comment (ochaplashkin) [reply test/autorequire_luatest_test.lua]: Fixed.

**Review by ylobankov — APPROVED** [review]
LGTM
